src/compiler/checker.ts

```diff
@@ -3643,9 +3630,17 @@ module ts {
         }
 
         function getReducedTypeOfUnionType(type: UnionType): Type {
-            // If union type was created without subtype reduction, perform the deferred reduction now
+            // If union type was created without subtype reduction, perform the deferred reduction now. If a circularity
+            // is detected, simply use the type itself.
             if (!type.reducedType) {
-                type.reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                type.reducedType = resolvingType;

```

Comment:
I don't get it, we always use `resolvingType` to detect other circularities, why would we be different here?